src/libraries/Common/src/System/Net/Http/aspnetcore/Http3/QPack/QPackEncoder.cs

```diff
@@ -222,18 +265,35 @@ public static bool EncodeValueString(ReadOnlySpan<string> values, string? separa
                     buffer = buffer.Slice(nameLength);
                     if (buffer.Length >= valueLength)
                     {
-                        string value = values[0];
-                        EncodeValueStringPart(value, buffer);
-                        buffer = buffer.Slice(value.Length);
-
-                        for (int i = 1; i < values.Length; ++i)
+                        if (valueEncoding is null)
                         {
-                            EncodeValueStringPart(separator, buffer);
-                            buffer = buffer.Slice(separator.Length);
-
-                            value = values[i];
-                            EncodeValueStringPart(value, buffer);
-                            buffer = buffer.Slice(value.Length);
+                            for (int i = 0; i < values.Length; i++)
+                            {
+                                if (i != 0)
+                                {
+                                    EncodeValueStringPart(separator, buffer);
+                                    buffer = buffer.Slice(separator.Length);
+                                }
```

Comment:
you've added a branch here that was previously optimized out with `for(int i = 1; ...)`

Reply:
Is the case of a single header value more common for Http3 than Http2? I am happy to revert this change, but I assume we would want the same approach for both Http2 and Http3 since the mehtods are almost the same.

The other difference I saw was when we calculate the encoded length, we use checked arithmetic for Http2, but not Http3? Should it be checked for both?

Reply:
Changed to having the `i == 0` outside the loop